File: src/main/java/example/websocket/client/WebSocketClient.java
package example.websocket.client;


import io.netty.bootstrap.Bootstrap;
import io.netty.buffer.Unpooled;
import io.netty.channel.Channel;
import io.netty.channel.ChannelInitializer;
import io.netty.channel.ChannelPipeline;
import io.netty.channel.socket.SocketChannel;
import io.netty.channel.socket.nio.NioEventLoop;
import io.netty.channel.socket.nio.NioSocketChannel;
import io.netty.handler.codec.http.HttpRequestEncoder;
import io.netty.handler.codec.http.HttpResponseDecoder;
import io.netty.handler.codec.http.websocketx.CloseWebSocketFrame;
import io.netty.handler.codec.http.websocketx.PingWebSocketFrame;
import io.netty.handler.codec.http.websocketx.TextWebSocketFrame;
import io.netty.handler.codec.http.websocketx.WebSocketClientHandshaker;
import io.netty.handler.codec.http.websocketx.WebSocketClientHandshakerFactory;
import io.netty.handler.codec.http.websocketx.WebSocketVersion;

import java.net.URI;
import java.util.HashMap;

public class WebSocketClient {

    private final URI uri;

    public WebSocketClient(URI uri) {
        this.uri = uri;
    }

    public void run() throws Exception {
        Bootstrap b = new Bootstrap();
        try {

            String protocol = uri.getScheme();
            if (!protocol.equals("ws")) {
                throw new IllegalArgumentException("Unsupported protocol: " + protocol);
            }

            HashMap<String, String> customHeaders = new HashMap<String, String>();
            customHeaders.put("MyHeader", "MyValue");

            // Connect with V13 (RFC 6455 aka HyBi-17). You can change it to V08 or V00.
            // If you change it to V00, ping is not supported and remember to change
            // HttpResponseDecoder to WebSocketHttpResponseDecoder in the pipeline.
            final WebSocketClientHandshaker handshaker =
                    new WebSocketClientHandshakerFactory().newHandshaker(
                            uri, WebSocketVersion.V13, null, false, customHeaders);

            b.eventLoop(new NioEventLoop())
             .channel(new NioSocketChannel())
             .remoteAddress(uri.getHost(), uri.getPort())
             .handler(new ChannelInitializer<SocketChannel>() {
                 @Override
                 public void initChannel(SocketChannel ch) throws Exception {
                     ChannelPipeline pipeline = ch.pipeline();
                     pipeline.addLast("decoder", new HttpResponseDecoder());
                     pipeline.addLast("encoder", new HttpRequestEncoder());
                     pipeline.addLast("ws-handler", new WebSocketClientHandler(handshaker));
                 }
             });

            System.out.println("WebSocket Client connecting");
            Channel ch = b.connect().sync().channel();
            handshaker.handshake(ch).sync();

            // Send 10 messages and wait for responses
            System.out.println("WebSocket Client sending message");
            for (int i = 0; i < 1000; i++) {
                ch.write(new TextWebSocketFrame("Message #" + i));
            }

            // Ping
            System.out.println("WebSocket Client sending ping");
            ch.write(new PingWebSocketFrame(Unpooled.copiedBuffer(new byte[]{1, 2, 3, 4, 5, 6})));

            // Close
            System.out.println("WebSocket Client sending close");
            ch.write(new CloseWebSocketFrame());

            // WebSocketClientHandler will close the connection when the server
            // responds to the CloseWebSocketFrame.
            ch.closeFuture().sync();
        } finally {
            b.shutdown();
        }
    }

    public static void main(String[] args) throws Exception {
        URI uri;
        if (args.length > 0) {
            uri = new URI(args[0]);
        } else {
            uri = new URI("ws://localhost:8080/sample.websocket");
        }
        new WebSocketClient(uri).run();
    }
}

File: src/main/scala/info/crlog/higgs/agents/websocket/client/WebSocketClient.scala
package info.crlog.higgs.agents.websocket.client

/**
 * Courtney Robinson <courtney@crlog.info>
 */

import io.netty.bootstrap.Bootstrap
import io.netty.channel._
import io.netty.channel.socket.SocketChannel
import io.netty.channel.socket.nio.NioEventLoopGroup
import io.netty.channel.socket.nio.NioSocketChannel
import io.netty.handler.codec.http.HttpRequestEncoder
import io.netty.handler.codec.http.HttpResponseDecoder
import io.netty.handler.codec.http.websocketx.{WebSocketClientHandshakerFactory, WebSocketVersion}
import java.util.HashMap
import java.net.URI
import info.crlog.higgs.agents.msgpack.{Interaction, Packing}
import scala.collection._
import io.netty.logging.{Slf4JLoggerFactory, InternalLoggerFactory, InternalLogger}
import info.crlog.higgs.agents.msgpack.commands.Subscribe

class WebSocketClient(host: String, port: Int) {
  InternalLoggerFactory.setDefaultFactory(new Slf4JLoggerFactory())
  private val log: InternalLogger = InternalLoggerFactory.getInstance(getClass)
  val listeners = mutable.Map.empty[Class[Any], (Any) => Unit]
  var channel: Channel = null

  /**
   * subscribes the given function to any messages of the type provided
   * The client must be connected to subscribe.
   * @param clazz   the type of classes the given function will be sent
   * @param fn the function that will be invoke for each message of the given type that is received
   * @tparam T Any subclass of the base Interaction class
   */
  def listen[T <: Interaction](clazz: Class[T], fn: (T) => Unit) {
    log.info("Subscribing for interactions of type: %s".format(clazz.getName))
    isConnected()
    val t = clazz.asInstanceOf[Class[Any]] -> fn.asInstanceOf[(Any) => Unit]
    listeners += t
    //subscribe the client to receive classes of this type
    send(new Subscribe(clazz.getName))
  }

  /**
   * Send a message to the server
   * @param msg the message to be sent
   * @tparam T
   * @return   the client to support chaining
   */
  def send[T <: Interaction](msg: T) = {
    isConnected
    if (!channel.isOpen) {
      throw new IllegalStateException("Connection to the server is no longer open")
    }
    channel.write(Packing.pack(msg))
    this
  }


  protected def isConnected() {
    if (channel == null) {
      throw new IllegalStateException("This client is not connected")
    }
  }

  def connect(onconnected: (ChannelFuture) => Unit) {
    log.info("Connecting to WebSocket server on %s:%s".format(host, port))
    val b: Bootstrap = new Bootstrap
    val customHeaders: HashMap[String, String] = new HashMap[String, String]
    val handshaker = new WebSocketClientHandshakerFactory()
      .newHandshaker(new URI("ws://" + host + ":" + port + "/"), WebSocketVersion.V13, null, true, customHeaders)
    b.group(new NioEventLoopGroup)
      .channel(new NioSocketChannel)
      .remoteAddress(host, port)
      .handler(new ChannelInitializer[SocketChannel] {
      def initChannel(ch: SocketChannel) {
        val pipeline: ChannelPipeline = ch.pipeline
        pipeline.addLast("decoder", new HttpResponseDecoder)
        pipeline.addLast("encoder", new HttpRequestEncoder)
        pipeline.addLast("ws-handler", new WebSocketClientHandler(handshaker, listeners))
      }
    })
    b.connect().addListener(new ChannelFutureListener {
      def operationComplete(future: ChannelFuture) {
        if (!future.isSuccess) {
          log.error("Failed to connect", future.cause())
        }
        channel = future.channel()
        log.info("Connected to Websocket server and channel is open:%s\nPerforming handshake".format(channel.isOpen))
        handshaker.handshake(channel).addListener(new ChannelFutureListener {
          def operationComplete(future: ChannelFuture) {
            log.info("Handshake complete and channel is open:%s".format(channel.isOpen))
            onconnected(future)
          }
        })
      }
    })
  }
}

